idx/memory/tag_query.go

```diff
@@ -0,0 +1,408 @@
+
+	// choose key-value combo that has the smallest number of series IDs because
+	// we want to start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.equal {
+		l := len(index[q.equal[i].key][q.equal[i].value])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// copy the map, because we'll later delete items from it
+	for k, v := range index[q.equal[startId].key][q.equal[startId].value] {
+		resultSet[k] = v
+	}
+
+	return startId, resultSet
+}
+
+func (q *TagQuery) filterByEqual(skipEqual int, resultSet TagIDs, index TagIndex) {
+	for i, e := range q.equal {
+		if i == skipEqual {
+			continue
+		}
+
+		for id := range resultSet {
+			if _, ok := index[e.key][e.value][id]; !ok {
+				delete(resultSet, id)
+			}
+		}
+	}
+}
+
+func (q *TagQuery) filterByNotEqual(resultSet TagIDs, index TagIndex, byId map[string]*idx.Archive) {
+	for _, e := range q.notEqual {
+		fullTag := e.key + "=" + e.value
+	IDS:
+		for id := range resultSet {
+			var def *idx.Archive
+			var ok bool
+			if def, ok = byId[id.ToString()]; !ok {
+				// should never happen because every ID in the tag index
+				// must be present in the byId lookup table
+				CorruptIndex.Inc()
+				delete(resultSet, id)
+				continue IDS
+			}
+			for _, tag := range def.Tags {
+				if tag == fullTag {
+					delete(resultSet, id)
+					continue IDS
+				}
+			}
+		}
+	}
+}
+
+// filters a list of metric ids by the expressions give. it is assumed that the
+// given expressions are all based on regular expressions.
+//
+// expressions: the list of key & value pairs
+// skipMatch:   an integer specifying an expression index that should be skipped
+// resultSet:   list of series IDs that should be filtered
+// byId:        ID keyed index of metric definitions, used to lookup the tags of IDs
+// not:         whether the resultSet shall be filtered by the =~ or !=~ condition
+//
+func (q *TagQuery) filterByMatch(expressions []kv, skipMatch int, resultSet TagIDs, byId map[string]*idx.Archive, not bool) error {
+	for i, e := range expressions {
+		if i == skipMatch {
+			continue
+		}
+
+		var shortCut bool
+		var re *regexp.Regexp
+		var err error
+
+		// shortcut if pattern is ^.+, any value will match so there's no need
+		// to actually run the regular expression.
+		// (f.e. expression "key!=" will be translated to "key=~^.+")
+		if e.value == "^.+" {
+			shortCut = true
+		} else {
+			re, err = regexp.Compile(e.value)
+			if err != nil {
+				return errInvalidQuery
+			}
+		}
+
+		// cache all tags that have once matched the regular expression.
+		// this is based on the assumption that many matching tags will be repeated
+		// over multiple series, so there's no need to run the regex for each of them
+		// because once we know that a tag matches we can just compare strings
+		matchingTags := make(map[string]struct{})
```

Comment:
Couldn't you equally cache nonMatches by making this `map[string]bool`? Then in the loop below you can delete if `matchingTags[tag] != not`.

This would also allow you to cache when the tagSplits[0] doesn't match the key that you are looking to compare against so you don't have to split it.

Reply:
I'm actually starting to doubt that this is a good idea, since the cache could grow quite large for high cardinality tags, and just growing the cache might become more expensive than running the regex. 

Of course, this is uncorroborated and just a gut feeling.

Reply:
true... i guess the cheap middle way would be to just limit the cache to some configurable value like `1000` items or whatever allows it to remain fast. if somebody has huge cardinality it would just stop caching after that count has been reached.

Reply:
assuming that many queries are going to be executed many times (f.e. if there's a dashboard that keeps refreshing every X seconds) we could even consider caching the regex matches and not matches across requests. but then we'd need some smart LRU mechanism to evict those cached matches and non-matches that haven't been used for a while. might also be over kill

Reply:
Implemented it and defined a default cache size limit of `1000` for now: https://github.com/grafana/metrictank/pull/729/commits/90c16f47533bb37aedfdb9c2b20c53cf1ebc7c81
Also added another benchmark that's very regex heavy: https://github.com/grafana/metrictank/pull/729/commits/ddd95a66c97040be6c6aabbb4dd0406264b9ee8e

There is quite an improvement:
```
before:
BenchmarkTagQueryFilterAndIntersectOnlyRegex-8   	      30	 593699132 ns/op	72545268 B/op	 2171554 allocs/op

after:
BenchmarkTagQueryFilterAndIntersectOnlyRegex-8   	      50	 434684230 ns/op	34478241 B/op	  976783 allocs/op
```

Reply:
Nifty. What happens if you make the cache very large?

Reply:
well, with the current benchmarks it won't make a difference since the limit doesn't get reached anyway.
the main consideration behind this was that at least in our current prod environment we seem to be more memory bound than cpu bound, so i don't want to let this grow endlessly.
in a hypothetical environment where there is unlimited memory i guess at some point the map lookup would be slower than the regex, but that would require the map to be really huge (i'm just guessing here).